### PR TITLE
feat(bus): C-102 — surface-router primitive + payload-filter (MIG-1.9, G-1111.A)

### PR DESCRIPTION
**Pull request by mellanon**
## What

MIG-1.9 — implement the **surface-router primitive** + **payload-filter** in `src/bus/`. New code per G-1111.A spec (read from `feat/g-1111-event-taxonomy` design doc §4–§5). This is the keystone for MIG-3b/4b/5b: those phases register adapters/runner/taps with this router.

## Components delivered

**`src/bus/payload-filter.ts`** (185 LOC):
- EventBridge content-filter grammar subset per G-1111 §4.3
- Operators: exact match (any-of), `anything-but`, `prefix`, `exists`, `equals-ignore-case`
- Pattern semantics: AND across keys, OR within a leaf array
- Filters apply to envelope context AND payload, nested by JSON structure
- Unknown operators fail closed (defensive)

**`src/bus/surface-router.ts`** (302 LOC):
- `SurfaceAdapter` interface (id, subjects, filter?, render, optional health)
- `SurfaceRouter`: register / unregister / start / stop / dispatch
- `subjectMatches(pattern, subject)` exported helper (NATS-style: `>` terminal, `*` single, literal)
- Adapter isolation: per-render timeout (default 5000ms) + `Promise.allSettled` fan-out + `onAdapterError` observability hook
- `dispatch()` never throws — all adapter errors captured

## API shape

\`\`\`typescript
const router = createSurfaceRouter(myelinRuntime, {
  renderTimeoutMs: 5000,
  onAdapterError: (id, err) => observability.report(id, err),
});

router.register({
  id: "discord-grove",
  subjects: ["local.metafactory.review.>"],
  filter: { payload: { repo: ["grove", "grove-v2"] } },
  render: async (envelope) => { /* post to Discord */ },
});

await router.start();
// ... envelopes flow ...
await router.stop();
\`\`\`

## Tests (79 new)

- **payload-filter** — 43 tests across 9 describe blocks: degenerate cases, exact-match, anything-but (scalar + array, missing-field semantics), prefix, exists (true/false × present/absent), equals-ignore-case, multi-key AND / multi-value OR, two-deep nesting, envelope-level matching, unknown-operator fail-closed.
- **surface-router** — 36 tests across 8 describe blocks: subjectMatches edge cases, register/unregister, subject-match dispatch (positive/negative/union), filter dispatch, adapter isolation (fail-isolated, throwing observer, sync-throw), timeout cancellation, lifecycle (start/stop idempotent, dispatch-after-stop no-op), runtime-integration shape.

## Verification

- cortex root \`bunx tsc --noEmit\` → **exit 0**
- \`bun test src/bus/__tests__/payload-filter.test.ts\` → **43/43**
- \`bun test src/bus/__tests__/surface-router.test.ts\` → **36/36**
- \`bun test src/bus/\` → **152/152** (was 73; **+79 net new**)
- Full suite: **1640/1641** (1 pre-existing React shell test, same as MIG-2)

## Design decisions

1. **`dispatch(envelope, subject?)` two-arg signature.** Spec says \`dispatch(envelope: Envelope): Promise<void>\` (test-only). I added an optional second \`subject\` parameter — strictly additive, doesn't break the spec contract — because reconstructing the NATS subject from envelope alone is unreliable (`envelope.type = domain.entity.action`; org segment isn't always derivable from `envelope.source`). When omitted (test path), `envelope.type` is the synthetic subject. Test-pinned.
2. **Defensive `console.error` fallback when `onAdapterError` not provided.** Drop-on-the-floor is unsafe for an isolation-critical path. Logs with `grove-bot: surface-router adapter "<id>" render failed:` prefix matching existing runtime/subscriber convention.
3. **Hanging render leaks until completion.** JS has no Promise cancellation; `Promise.race` resolves but the loser keeps running. Acceptable v1 (dispatch unblocked, which is what §5.3 requires). Future: pass `AbortSignal` to `render()`.
4. **`MyelinRuntime` integration deferred to follow-up PR.** Current runtime hardcodes `logEnvelope` as the per-envelope handler — no public registration seam. Per task contract I did NOT modify the runtime in this PR. The router accepts runtime as constructor param to preserve the contract. Wiring is a one-line edit at `runtime.ts:95` in a follow-up: add `onEnvelope(handler)` registration; existing logging keeps working alongside fan-out to router. Once that lands, MIG-1.9 plan-tick + the §5.3 integration acceptance test become natural.

## What this PR is NOT

- **No MyelinRuntime modification** — explicitly out of scope per task contract; lands in follow-up.
- **No YAML adapter-config-loader** — spec §4.4. Adapters can be programmatically registered from cortex.ts entrypoint (MIG-7.1); YAML config loader is G-1111.E follow-up.
- **No specific adapter implementations** — Discord/Mattermost/PagerDuty/Dashboard adapter classes land in MIG-3b/4b plus future work.

## Plan grounding

`docs/architecture.md` §8 — bus/ at top level. Plan §4 MIG-1.9. Spec at `/tmp/g-1111-spec.md` (saved from `feat/g-1111-event-taxonomy:docs/design-event-taxonomy.md`) §4 + §5.

Refs cortex#3 (MIG-1 umbrella).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/bus/surface-router.ts [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+ *   exposes no public registration point for an external consumer like
+ *   this router. Wiring the router to receive live envelopes therefore
+ *   requires a runtime change to add an `onEnvelope(handler)` registration
+ *   surface, which is explicitly out of scope for this PR (per the task
+ *   contract: "DO NOT modify MyelinRuntime in this PR").
+ *
+ *   This file ships the router as a self-contained dispatcher whose
+ *   public `dispatch(envelope, subject?)` is the integration entry point.
+ *   Tests exercise it directly. A follow-up PR will add the runtime hook
+ *   so production envelopes flow into `dispatch()`.
```
Comment (mellanon) [inline src/bus/surface-router.ts:27]: **[advisory/architecture]** Per the migration plan (`docs/plan-cortex-migration.md` §4.345), MIG-1.9 reads as "surface-router subscribes to NATS via `MyelinRuntime` and fans envelopes" — and the §4.355 acceptance test is "starting `MyelinRuntime`, registering one fake adapter with the surface-router, publishing one envelope, asserting the adapter received it." This PR ships the dispatcher only; the wiring is deferred. The rationale in the PR body is defensible (cleaner review unit, runtime API change has its own risk), but the primitive sits **unused in production** until the wiring follow-up lands.

This is acceptable IF the follow-up PR is opened immediately so the integration test can land before MIG-3b/4b/5b need to register adapters. Recommend: open the runtime-wiring PR within 24h of this merge, or call out the wiring in this PR as scope creep that you have explicit authorization to defer.

Not a blocker for this PR — the dispatcher is correct in isolation and the deferred wiring is one localised edit. But the team should be aware that the §4.345/§4.355 acceptance criteria are NOT satisfied by this PR alone.

File: src/bus/surface-router.ts [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+   *  hook, `start()` will be where we register that. */
+  start(): Promise<void>;
+  /** Stop accepting new dispatches. After stop, `dispatch()` is a no-op
+   *  (still returns a resolved promise). Idempotent. */
+  stop(): Promise<void>;
+  /** Dispatch one envelope to all matching adapters. Subject is optional;
+   *  when omitted, it is derived from `envelope.type` for local testing.
+   *  Production callers (the future runtime hook) MUST pass the actual
+   *  NATS subject so wildcard patterns match correctly. */
+  dispatch(envelope: Envelope, subject?: string): Promise<void>;
```
Comment (mellanon) [inline src/bus/surface-router.ts:88]: **[advisory/architecture]** The optional second `subject` argument is a defensible deviation from the spec`s `dispatch(envelope: Envelope): Promise<void>` shape. Confirmed:

- `envelope.type` per G-1111 §3.1 naming rules = `domain.entity.action` (e.g. `review.cycle.completed`)
- The NATS subject = `local.{org}.{type}` (e.g. `local.metafactory.review.cycle.completed`)
- The org segment is NOT derivable from `envelope.source` (DID-shaped, not a subject prefix)

Without the second arg, `subjectMatches("local.metafactory.review.>", envelope.type)` would silently fail to match because `envelope.type` starts with `review.`, not `local.metafactory.review.`. The two-arg signature is the **correct fix** — the synthetic-subject test path is acceptable because the test uses `subjects: [envelope.type]`, sidestepping the `local.{org}.` prefix. Production callers (the future runtime hook) MUST pass the actual NATS subject, which the JSDoc on line 86–88 correctly states.

Suggestion: when the runtime wiring lands, add a comment-test that asserts a misuse case — passing only `envelope.type` to a router with `local.{org}.>` patterns produces zero matches. This pins the contract: "subject is required for production NATS patterns; envelope.type is test-only convenience."

File: src/bus/surface-router.ts [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+      timer = setTimeout(() => {
+        reject(new Error(`render timeout after ${timeoutMs}ms`));
+      }, timeoutMs);
+    });
+
+    // Defensively wrap the render call in case it throws synchronously
+    // before returning a promise.
+    const renderPromise = (async () => adapter.render(envelope))();
+
+    await Promise.race([renderPromise, timeoutPromise]);
```
Comment (mellanon) [inline src/bus/surface-router.ts:270]: **[advisory/hardening]** The hanging-render leak (loser of `Promise.race` keeps running) is correctly identified as a v1 limitation in the PR body. §5.3 ("MUST NOT block others") is satisfied — `dispatch()` unblocks at `timeoutMs`. The leak is bounded by `concurrent-renders × adapters × bytes-per-render`.

Two follow-up notes:

1. **`AbortSignal` is the right v2.** When you wire it, the API change is `render(envelope: Envelope, signal: AbortSignal): Promise<void>` — adapters opt-in to honoring it (Discord posters, fetch calls, etc.). Adapters that ignore the signal still have their result discarded; the leak just becomes a render-completion-after-discard rather than render-keeps-running-forever. Net win.

2. **Operational telemetry.** Right now the `safeReportError` reports the `Error` but not which kind (timeout vs. throw). Consider `err.message.startsWith("render timeout")` as a discriminator, or — better — wrap timeout errors in a typed `RenderTimeoutError extends Error` so observers can distinguish `system.adapter.timeout` events (per §3.5 vocabulary) from `system.adapter.render-failed`.

Neither is a blocker. Both belong in the §3.5 system-event publisher work (likely G-1111.B/D).

File: src/bus/surface-router.ts [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+function safeReportError(
+  adapterId: string,
+  err: Error,
+  hook: SurfaceRouterOptions["onAdapterError"],
+): void {
+  if (!hook) {
+    // Default sink: log so a missing observer doesn't make adapter
+    // failures invisible.
+    console.error(
+      `grove-bot: surface-router adapter "${adapterId}" render failed:`,
```
Comment (mellanon) [inline src/bus/surface-router.ts:288]: **[suggestion/quality]** Defensive `console.error` fallback when `onAdapterError` is not provided is the right v1 default — silent drop on an isolation-critical path would hide adapter failures from operators. Format `grove-bot: surface-router adapter "<id>" render failed:` matches sibling convention in `src/bus/myelin/runtime.ts` and `subscriber.ts`.

One minor: when this primitive lives in `cortex` (not `grove-bot`), the prefix `grove-bot:` will read oddly. Recommend a follow-up to standardise log prefixes across `src/bus/` to either `cortex:` or a per-component prefix (`surface-router:`, `myelin-runtime:`, etc.) — but that is repository-wide and out of scope for this PR.

File: src/bus/payload-filter.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    return (
+      typeof actual === "string" &&
+      actual.toLocaleLowerCase() === candidate["equals-ignore-case"].toLocaleLowerCase()
+    );
+  }
+
+  // Unknown operator — fail closed. New operators must be explicitly
+  // added here; silently accepting an unknown shape would mask config
+  // typos.
+  return false;
```
Comment (mellanon) [inline src/bus/payload-filter.ts:180]: **[advisory/quality]** Unknown-operator fail-closed at line 180 is exactly right — silent acceptance of malformed YAML would mask config typos AND open a degenerate matching path (`{wildcard-glob: "*"}` returning `true` because no key matched would be a security incident). The fail-closed default is documented in the JSDoc and tested explicitly at `payload-filter.test.ts:370-381`.

Tiny suggestion: when YAML adapter-config-loader (§4.4 / G-1111.E) lands, surface the failed match more loudly than just `false` — log a warning like `payload-filter: unknown operator "wildcard-glob" in adapter "<id>" filter — failing closed`. Right now a typo silently filters all envelopes out, which presents as "my Discord adapter stopped working" rather than "you have a typo in `~/.config/grove/surfaces/discord-grove.yaml`." But that diagnostic improvement belongs at config-load time, not in the matcher itself. Defer to G-1111.E.

File: src/bus/surface-router.ts [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+ * future wiring hook (see file header) — the router does not call into
+ * the runtime today; the runtime will call into the router via
+ * `dispatch()` once an `onEnvelope` registration surface lands.
+ */
+export function createSurfaceRouter(
+  // Reserved for the future runtime → router wiring (see file header).
+  // Today the parameter is part of the contract but the router does not
+  // call into the runtime; the runtime will call `dispatch()` on this
+  // router once an `onEnvelope` registration surface is added there.
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (mellanon) [inline src/bus/surface-router.ts:169]: **[nit/quality]** The `eslint-disable-next-line @typescript-eslint/no-unused-vars` is redundant — the parameter is already underscore-prefixed (`_runtime`), which by convention satisfies the unused-vars rule in most TypeScript ESLint configs. If the project ESLint config has `argsIgnorePattern: "^_"` (the standard TS-ESLint preset), the disable is no-op; if it doesn`t, the disable is masking a configuration that should be tightened.

Either remove the `eslint-disable-next-line` and rely on `_` prefix convention (preferred), OR remove the `_` prefix and keep the disable (worse — discloses the intent less clearly). Not a blocker.

**Comment by mellanon** [comment]
## Full Review Summary: cortex#25

**Verdict: APPROVE — recommend merge.** The first cortex new-code PR is solid. Spec fidelity is excellent, all isolation invariants hold, 79 new tests cover every operator + every edge case I checked. Five advisory comments posted; none are blockers.

### Lens Results

| Lens | Critical | Warning | Suggestion | Nit | Verdict |
|------|----------|---------|------------|-----|---------|
| CodeQuality | 0 | 0 | 1 | 1 | pass |
| Security | 0 | 0 | 0 | 0 | pass |
| Hardening | 0 | 0 | 1 | 0 | pass |
| Architecture | 0 | 0 | 2 | 0 | pass (advisory) |
| EcosystemCompliance | 0 | 0 | 0 | 0 | pass |
| Performance | 0 | 0 | 0 | 0 | pass |
| Duplication | 0 | 0 | 0 | 0 | pass |

### Spec-fidelity audit (G-1111 §4 + §5)

- §4.1 `SurfaceAdapter` interface: `id, subjects, filter?, render, health?` — exact match
- §4.2 Subject patterns: `>` terminal (`sLen >= pLen` correct) ✓; `>` non-terminal returns false (defensive) ✓; `*` single segment, no spanning ✓; empty pattern/subject no-match ✓; `>` alone matches any non-empty ✓
- §4.3 All 5 v1 operators: exact (any-of), anything-but (scalar + array), prefix, exists (true/false), equals-ignore-case ✓
- §4.3 "AND across keys, OR within a leaf array" semantics ✓ (`matchPattern` walks all keys; `matchLeaf` short-circuits on first match)
- §4.3 Filters apply to envelope context AND payload ✓ (`filter.envelope` + `filter.payload`)
- §4.3 Missing field semantics: `exists:false` = absent predicate; `anything-but` requires presence (matches EventBridge); `prefix`/`equals-ignore-case` fail closed on type mismatch ✓
- §5.1 Single subscription source — router does NOT open its own subscriptions ✓ (consistent with deferred wiring)
- §5.3 Adapter isolation — `Promise.allSettled` + per-render timeout + tests prove (a) failing adapter doesn't block, (b) sync throws captured, (c) throwing observer can't poison, (d) timeout cancels at boundary, (e) `dispatch()` never throws

### On the design decisions

1. **`dispatch(envelope, subject?)` two-arg deviation.** **Endorsed.** `envelope.type` (`review.cycle.completed`) is NOT the NATS subject (`local.{org}.review.cycle.completed`); the org segment can't be derived from `envelope.source` reliably. Strictly additive, doesn't break the spec contract. Production callers MUST pass the subject (JSDoc on lines 86-88 says so).
2. **Defensive `console.error` fallback.** **Endorsed.** Drop-on-the-floor would be unsafe for an isolation-critical path. Format matches sibling convention.
3. **Hanging render leak.** **Acceptable v1 limitation.** §5.3 satisfied (dispatch unblocks). Memory leak bounded by `adapters × throughput`. `AbortSignal` is the right v2.
4. **Deferred MyelinRuntime wiring.** **Acceptable but flagged.** Plan §4.345/§4.355 acceptance criteria explicitly call for the router to "subscribe to NATS via `MyelinRuntime`" — this PR ships the dispatcher only. The author's rationale (cleaner review unit, isolated runtime API change) is defensible IF the follow-up PR opens immediately. The risk: this primitive sits unused until the wiring follow-up lands. Recommend opening the runtime-wiring PR within 24h of this merge.

### Spot-checks performed (worktree at ~/Developer/cortex-mig19-surface-router/)

- `bunx tsc --noEmit` → exit 0 ✓
- `bun test src/bus/__tests__/payload-filter.test.ts` → **43/43** ✓
- `bun test src/bus/__tests__/surface-router.test.ts` → **36/36** ✓
- `bun test` (full) → **1640 pass / 1 pre-existing fail** (jsonl-reader malformed-line warning only — same as MIG-2)
- `gh pr view 25 --json mergeable` → **MERGEABLE / CLEAN**
- Verified no duplicate `subjectMatches` exists elsewhere in cortex (grep across `src/`)
- Verified `runtime.ts:95` hard-codes `logEnvelope` as the per-envelope handler (confirms author's stated wiring point)
- Verified `MyelinSubscriber.start({pattern, onEnvelope})` already takes a per-envelope callback — the runtime-wiring follow-up is genuinely a one-line plumbing edit

### Key observations

- **File headers are exemplary.** Both files explicitly cite spec sections (§4.3, §5, §5.1, §5.3) and explain deviations with rationale. This is the standard the rest of cortex should match.
- **Test coverage is thorough.** 79 new tests, each `describe` block maps to a spec property. The unknown-operator fail-closed test (line 370-381) is a great example of pinning a security-critical defensive choice.
- **The `_runtime` placeholder is honest.** The constructor takes `MyelinRuntime` but doesn't use it today — the file header explains why explicitly. This is better than fake-using it or refactoring the API to pretend it's not needed.
- **No duplication.** Subject-matching is a fresh implementation; cortex doesn't have a prior NATS pattern matcher to consolidate with. NATS does broker-side matching; this client-side matcher is the first cortex needs.

### Recommendation

**APPROVE — recommend merge.** This is a solid foundation for MIG-3b/4b/5b. The deferred runtime wiring is the only architectural risk and is defensible. Open the wiring follow-up PR immediately so the §4.355 integration acceptance test can land before adapters need to register.

verdict: blockers=0 majors=0 nits=5 — recommend: merge

**Review by mellanon — COMMENTED** [review]
Full review (6 lenses + duplication) — first cortex new-code PR. Spec-fidelity audit against G-1111 §4 + §5: all 5 payload-filter operators correct; `>` terminal/`*` single-segment subject matching correct on every edge case I checked; adapter isolation invariants (failing adapter doesn't block, sync-throw captured, throwing observer can't poison, timeout cancels, dispatch never throws) all proven by tests. 79 new tests, 1640/1641 full suite (1 pre-existing fail), `bunx tsc --noEmit` clean. Five advisory inline comments — none blocking. Deferred MyelinRuntime wiring is defensible but means plan §4.345/§4.355 acceptance criteria are NOT satisfied by this PR alone — open the runtime-wiring follow-up PR immediately so the integration test can land before MIG-3b/4b/5b need it.

verdict: blockers=0 majors=0 nits=5 — recommend: merge (posted as comment-review — bot account opened the PR; --approve blocked by GitHub)
